src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs

```diff
@@ -0,0 +1,194 @@
+﻿// ***********************************************************************
+// Copyright (c) 2017 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+#if PARALLEL
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using NUnit.Framework.Interfaces;
+using NUnit.TestData.ParallelExecutionData;
+using NUnit.TestUtilities;
+
+namespace NUnit.Framework.Internal.Execution
+{
+    [TestFixtureSource(nameof(GetParallelSuites)), Explicit]
+    public class ParallelExecutionTests : ITestListener
+    {
+        private TestSuite _testSuite;
+        private int _numCases;
+
+        private List<string> _events;
+        private TestResult _result;
+
+        private static readonly string NL = Environment.NewLine;
+
+        public ParallelExecutionTests(TestSuite testSuite, int numCases)
+        {
+            _testSuite = testSuite;
+            _numCases = numCases;
+        }
+
+        [OneTimeSetUp]
+        public void RunTestSuite()
+        {
+            _events = new List<string>();
+
+            var dispatcher = new ParallelWorkItemDispatcher(4);
+            var context = new TestExecutionContext();
+            context.Dispatcher = dispatcher;
+            context.Listener = this;
+
+            dispatcher.ShiftStarting += (shift) =>
+            {
+                AddEvent("ShiftStarted " + shift.Name);
+            };
+
+            dispatcher.ShiftFinished += (shift) =>
+            {
+                AddEvent("ShiftFinished " + shift.Name);
+            };
+
+            var workItem = TestBuilder.CreateWorkItem(_testSuite, context);
+
+            var run_complete = new ManualResetEvent(false);
+            workItem.Completed += (s, e) => { run_complete.Set(); };
```

Comment:
(`(s, e) => run_complete.Set();` is an option too)

Reply:
Also `run_complete` should really be in a `using` statement because it holds unmanaged handles.